src/algorithms.hpp

```diff
@@ -118,19 +119,30 @@ bool is_minimal_degree(C const& c)
     return it==c.end();
 }
 
-template <typename C>
-bool is_positive(C const& c)
-{
-    static_assert(
-        std::is_integral<typename C::value_type>::value,
-        "is_positive only applies to integral types"
-    );
-    for(auto v : c) {
-        if(v<1) {
-            return false;
-        }
+struct generic_is_positive {
+    template <typename V>
+    bool operator()(V v) const {
+        static V zero = V{};
+        return v>zero;
     }
-    return true;
+};
+
+struct generic_is_negative {
+    template <typename V>
+    bool operator()(V v) const {
+        static V zero = V{};
+        return v<zero;
+    }
+};
+
+template <typename C>
+bool all_positive(const C& c) {
+    return util::all_of(c, generic_is_positive{});
+}
+
+template <typename C>
+bool all_negative(const C& c) {
+    return util::all_of(c, generic_is_negative{});
```

Comment:
For consistent naming with what the standard uses, I guess something like the following would be appropriate,
```
template <typename V=void>
struct greater_than_zero {
    bool operator()(const V& v) const { return v>V{}; }
};

template <>
struct greater_than_zero<void> {
    template <typename V>
    bool operator()(const V& v) const { return v>V{}; }
};
```
where the default (void) case corresponds to the generic predicate. `is_foo<T>` should probable be kept for predicates on types rather than values.

What do you think?